pkg/reconciler/namespace/namespace.go

```diff
@@ -0,0 +1,165 @@
+/*
+Copyright 2019 The Knative Authors
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+	http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package namespace
+
+import (
+	"bytes"
+	"context"
+	"errors"
+	"regexp"
+	"text/template"
+
+	"go.uber.org/zap"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/equality"
+	apierrs "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	kubelisters "k8s.io/client-go/listers/core/v1"
+	"knative.dev/pkg/controller"
+	"knative.dev/pkg/logging"
+	"knative.dev/serving/pkg/apis/networking"
+	"knative.dev/serving/pkg/apis/networking/v1alpha1"
+	listers "knative.dev/serving/pkg/client/listers/networking/v1alpha1"
+	"knative.dev/serving/pkg/network"
+	"knative.dev/serving/pkg/reconciler"
+	"knative.dev/serving/pkg/reconciler/namespace/config"
+)
+
+// Reconciler implements controller.Reconciler for Certificate resources.
+type Reconciler struct {
+	*reconciler.Base
+
+	// listers index properties about resources
+	nsLister            kubelisters.NamespaceLister
+	knCertificateLister listers.CertificateLister
+
+	configStore configStore
+}
+
+// Check that our Reconciler implements controller.Reconciler
+var _ controller.Reconciler = (*Reconciler)(nil)
+
+// Reconciler implements controller.Reconciler for Namespace resources.
+func (c *Reconciler) Reconcile(ctx context.Context, key string) error {
+	logger := logging.FromContext(ctx)
+	ctx = c.configStore.ToContext(ctx)
+
+	original, err := c.nsLister.Get(key)
+
+	if apierrs.IsNotFound(err) {
+		logger.Errorf("Namespace %s in work queue no longer exists %s", key, err)
+		return nil
+	} else if err != nil {
+		return err
+	}
+
+	// Don't modify the informers copy
+	newNamespace := original.DeepCopy()
+
+	err = c.reconcile(ctx, newNamespace)
+	if err != nil {
+		c.Recorder.Event(newNamespace, corev1.EventTypeWarning, "InternalError", err.Error())
+	}
+	return err
+}
+
+func (c *Reconciler) reconcile(ctx context.Context, ns *corev1.Namespace) error {
+	cfg := config.FromContext(ctx)
+
+	// Only create wildcard certs for the default domain
+	defaultDomain := cfg.Domain.LookupDomainForLabels(nil)
```

Comment:
See my resource naming comment below, but perhaps how we ultimately address this should factor into that?

Perhaps we should name the resource `*.namespace.example.com`?

Reply:
This seems like an altogether better contract between controllers *like* this one and the Route controller.

Reply:
(also, I'm fine with this limitation for now, just trying to accommodate fixing this later)

Reply:
I don't follow about the contract between controllers. Can you elaborate?

Reply:
We have one controller that's responsible for creating wildcards certificates per namespace.

The Route controller is sensitive to wildcard certificates per-namespace.

Right now we're relying on fixed naming conventions between these, and I'm wondering whether that can be improved.

Reply:
I've changed this so that the route reconciler will look for an existing wildcard cert that matches before creating a cert.